File: ctci/java/ch1-arrays-and-strings/q5_one_away/Question.java
package q5_one_away;

/**
 * Created by jjlee on 7/11/16.
 */
public class Question {
    /* compare string lengths to use only one iteration */
    public static boolean oneEditAway(String s1, String s2) {
        if (Math.abs(s1.length() - s2.length()) > 1)
            return false;

        if (s1.length() < s2.length())
            return oneEditAwayHelper(s1, s2, false);
        else if (s1.length() > s2.length())
            return oneEditAwayHelper(s2, s1, false);
        else
            return oneEditAwayHelper(s1, s2, true);
    }

    /* helper fuction common for all three cases (insert, remove, replace) */
    private static boolean oneEditAwayHelper(String shorter, String longer, boolean forReplace) {
        boolean found = false;
        for (int i = 0; i < shorter.length(); i++) {
            if (shorter.charAt(i) != longer.charAt(!forReplace && found ? i + 1: i)) {
                if (found)
                    return false;
                found = true;
            }
        }
        return true;
    }
}

File: ctci/python/ch1-arrays-and-strings/q5-one-away/Question.py
# compare string lengths to use only one iteration
def one_edit_away(s1, s2):
    if abs(len(s1) - len(s2)) > 1:
        return False

    if len(s1) < len(s2):
        return one_edit_away_helper(s1, s2, False)
    elif len(s1) > len(s2):
        return one_edit_away_helper(s2, s1, False)
    else:
        return one_edit_away_helper(s1, s2, True)


# helper fuction common for all three cases (insert, remove, replace)
def one_edit_away_helper(shorter, longer, for_replace):
    found = False
    for i in xrange(len(shorter)):
        if shorter[i] != longer[i + 1 if not for_replace and found else i]:
            if found:
                return False
            found = True
    return True
